sprints/dashboard/tasks.py

```diff
@@ -202,11 +265,21 @@ def complete_sprint_task(board_id: int) -> None:
         )
         issue_keys = [issue.key for issue in issues]
 
+        cell_dict = {
+            'key': cell.key,
+            'name': cell.name,
+            'board_id': cell.board_id,
+        }
+
         # It is not mentioned in Python lib docs, but the limit for the issue-moving queries is 50 issues. Source:
         # https://developer.atlassian.com/cloud/jira/software/rest/#api-rest-agile-1-0-backlog-issue-post
         # https://developer.atlassian.com/cloud/jira/software/rest/#api-rest-agile-1-0-sprint-sprintId-issue-post
         batch_size = 50
         if not settings.DEBUG:  # We really don't want to trigger this in the dev environment.
+            # Raise error if we can't read roles from the handbook
+            get_cell_member_roles(raise_exception=True)
+            trigger_new_sprint_webhooks.delay(cell_dict, next_sprint.name, get_sprint_number(next_sprint), board_id)
```

Comment:
> Shouldn't we trigger this after ending the sprint?

I agree, the event should only be sent after ending the sprint, and the new sprint has already started.

> Will the receiving part have some safety checks that will verify the sprint number? 

The receiving party, for now, has no safety checks that will verify the sprint number.
We expect the receiving party to use the information without "second guessing" it.

Reply:
Sorry for sneaking in. @nizarmah I would say that the receiving party should use the information without second-guessing the incoming value **but** it **should** make sure that the sprint id is unique and not 0.

Reply:
No worries @gabor-boros 

> it should make sure that the sprint id is unique and not 0

But if the sprint id is 0 or not unique, the receiving party would simply not do anything about that. Instead, it would dismiss such a received event; please correct me if I'm wrong.

Based on that, there shouldn't be an issue/problem with trigger the event after ending the sprint.

---

Also the `get_cell_member_roles` and the `trigger_new_sprint_webhooks` don't have to be placed under the same condition statement. 

We want them to actually be placed in two different places in the method, so it would be better to separate them from under that conditional.

We do want to run the `get_cell_member_roles` though before ending the sprint, because we don't want to interrupt ending the sprint if we don't have any roles; instead, it would be better not to end the sprint at all, until we have the roles.

But the triggering of the event, that needs to be done at the end, after all the code that would be responsible for ending the sprint. And in case that's asynchronous, it should be done using a callback.

Reply:
@nizarmah From API point of view it would be expected to return 400 when an invalid data received, so - in my opinion - client should handle that scenario (and not crash).

One way of that could be to simply send an email to us about the fact the request failed.

Reply:
@gabor-boros but this should be a webhook event. So, any triggered/sent event should be a valid event, or the only source of truth. We're not waiting for a response, we're just posting data...

Crafty, a third party, should not be aware of what a valid Sprints "new sprint" event is... Crafty should only be aware of whether an event is useful for it or not. For example, let's say Crafty receives an event with no roles assigned to the members... Crafty should treat that event as the only source of truth and assign to the different members only tasks that are for Cell Members.

Now, sure, we might choose to add a condition so that Crafty doesn't create new checklists when the roles are empty, but that doesn't make the Sprints event any less valid in the eyes of Crafty, it just makes it inappropriate for its use.

So, for example, simply we'd have a condition that checks that the sprint id in the received event is valid. Otherwise, we'd **have Crafty ping** the _urgent_ mailing list.

If an invalid webhook event is triggered, then we haven't verified the data before sending an event. Any validation needs to be done in Sprints, and not in Crafty. If we return 400 and Sprints actually listens to the response, that means we haven't separated the services from each other, which is the main goal behind having Crafty.

And yes, I know that this is being said, while I'm requesting that we rely on the handbook to read the roles. And I know that we will use Crafty to create different checklists on monday.com. 😔 

But once we implement the open source solution, I have a bunch of work planned for Crafty, to get it closer to being a service (until we have an epic for it).